Graph synchro fix and copy/paste improvement.

hmiomand opened the pull request:
This PR fixes the issue of .diagram going out-of-sync with the .pi.
It should fix issues #351, #91 and #122.

This PR also fixes issues when copy/pasting with a Delay.
Delay getter/setter are properly handled.
However, Delay Dependencies are ignored as a stopgap solution.
An issue will regarding this will be created.

All tests passed, aside from RCPTT (see #352).

hmiomand committed:
Fix #351 about de-sync issues between .pi and .diagram files.
PiGraph vertex list is now updated in the same manner than the PiWriter writes .pi files.
The vertex list is filled with its 1st third for parameters, 2nd third for actors, and last third for delay.

hmiomand committed:
(UI) Follow up on #351, now handling properly addition/removal of fifo with/without delay and dependency in the .diagram to match the order in .pi.
Still have some occasional issues when fiddling around with delays

hmiomand committed:
(UI) Follow up on #351: indexes now handled in by the GraphInterfaceObserver (which will probably be renamed GraphObserver).
Delay actor are now forced to the end of the vertex list.

Might also fix #91 and #122.

hmiomand committed:
Fixed fifo index unable to be decremented

hmiomand committed:
Reordered operation in PiSDFSubgraphBuilder and PiSDFSubgraphBuilderTest for proper fifo addition and removal

hmiomand committed:
Added the method affectDelay in place of setDelay.
It ensures fifos stay properly registered and ordered in their graph.
setDelay should not be used.

hmiomand committed:
Fixed case where DelayActor would be added to PiGraph vertex list as a regular actor, instead of being added at the end

hmiomand committed:
Dirty fix for when the observer isn't triggered on a removeActor call

hmiomand committed:
Fixed uncatch exception when adding element to PiGraph vertex
Added safeguard when removing a PiGraph from a vertex list

hmiomand committed:
Added additional checks when adding/removing element from a PiGraph

hmiomand committed:
Added additional check when removing an actor or a delay from a PiGraph

hmiomand committed:
Refactored GraphInterfaceObserver into GraphObserver, and updated doc

hmiomand committed:
Fix issue when adding/removing a delay to a fifo which is already within a PiGraph.
The storage index and placement in the vertex list would not be updated accordingly.
Updated the GraphObserver to remove and re-add the fifo when a delay is added/removed.

hmiomand committed:
Renamed affectDelay function to assignDelay.

hmiomand committed:
Updated javadoc in PiSDF.xcore

hmiomand committed:
Added a PiMMUserFactory.copyWithHistory method (which calls on already existing PreesmUserFactory.copyWithHistory),
with properly check if the PiGraph, its subgraphs, and all concerned fifos, have a GraphObserver, adding one if missing.

hmiomand committed:
Adding the PiGraph.getAllChildrenGraphs() method missing from previous commit.

hmiomand committed:
Added a warning when an actor removal hasn't been detected by the Observer.

hmiomand committed:
Fixed subgraph removal from topgraph vertex list being done with remove() instead of removeActor()

hmiomand committed:
Fixed parameter removal from graph vertex list being done with remove() instead of removeParameter()

hmiomand committed:
Replaced getVertices() by the right specific getter

hmiomand committed:
Potential fix for copy/pasting groups of actor with a delay.
In case of delays with getters/setters, assiciated fifo are not yet pasted.

hmiomand committed:
Updated release notes

hmiomand committed:
Modified the delay copy/paste fix to be cleaner

hmiomand committed:
Fixed Delay setter/getter fifos not being pasted during a copy/paste

hmiomand committed:
Fixed condition for copy/pasting Delay

hmiomand committed:
Dirty fix to prevent tracking Delays Dependencies during pasting operation.
This allows the pasting operation with a Delay with Dependencies to complete instead of simply fail.
TODO : Fix the issue of dependencies on Delay while copy/pasting

Commit:
Merge branch 'develop' into graph_synchro_fix

ahonorat commented:
RCPTT tests succesfully passed on my machine (former installation not subject to issue #352 ).

hmiomand committed:
Only a single instance of GraphObserver is now in use.
The constructor in now private and the instance is fetched via a static method.
